wardrobe/tradable_kitties/src/tests.rs

```diff
@@ -0,0 +1,864 @@
+        Err(TradeableKittyError::NumberOfInputOutputMismatch)
+    );
+}
+
+#[test]
+fn list_for_sale_with_wrong_output_type_amoung_valid_output_fails() {
+    let input1 = TradableKittyData::default_kitty();
+    let mut output1 = TradableKittyData::default_tradable_kitty();
+    output1.kitty_basic_data = input1.clone();
+    let result = TradableKittyConstraintChecker::<0>::ListKittyForSale.check(
+        &[input1.into()],
+        &[],
+        &[Bogus.into()],
+    );
+    assert_eq!(result, Err(TradeableKittyError::BadlyTyped));
+}
+
+#[test]
+fn list_kitty_for_sale_with_wrong_input_type_fails() {
+    let result = TradableKittyConstraintChecker::<0>::ListKittyForSale.check(
+        &[Bogus.into()],
+        &[],
+        &[TradableKittyData::default_tradable_kitty().into()],
+    );
+    assert_eq!(result, Err(TradeableKittyError::BadlyTyped));
+}
+
+#[test]
+fn list_kitty_for_sale_with_input_missing_fails() {
+    let input1 = TradableKittyData::default_kitty();
+
+    let mut output1 = TradableKittyData::default_tradable_kitty();
+    output1.kitty_basic_data = input1.clone();
+    let mut output2 = TradableKittyData::default_tradable_kitty();
+    output2.kitty_basic_data.dna = KittyDNA(H256::from_slice(b"superkalifragislisticexpialadoca"));
+    let result = TradableKittyConstraintChecker::<0>::ListKittyForSale.check(
+        &[],
+        &[],
+        &[output1.into(), output2.into()],
+    );
+    assert_eq!(
+        result,
+        Err(TradeableKittyError::NumberOfInputOutputMismatch)
+    );
+}
+
+#[test]
+fn list_kitty_for_sale_with_duplicate_input_fails() {
+    let input1 = TradableKittyData::default_kitty();
+    let input2 = TradableKittyData::default_kitty();
+
+    let mut output1 = TradableKittyData::default_tradable_kitty();
+    output1.kitty_basic_data = input1.clone();
+    let mut output2 = TradableKittyData::default_tradable_kitty();
+    output2.kitty_basic_data.dna = KittyDNA(H256::from_slice(b"superkalifragislisticexpialadoca"));
+    let result = TradableKittyConstraintChecker::<0>::ListKittyForSale.check(
+        &[input1.into(), input2.into()],
+        &[],
+        &[output1.into(), output2.into()],
+    );
+    assert_eq!(result, Err(TradeableKittyError::DuplicateKittyFound));
+}
+
+#[test]
+fn list_kitty_for_sale_with_duplicate_output_fails() {
+    let input1 = TradableKittyData::default_kitty();
+    let input2 = TradableKittyData::default_kitty();
+
+    let mut output1 = TradableKittyData::default_tradable_kitty();
+    output1.kitty_basic_data = input1.clone();
+    let mut output2 = TradableKittyData::default_tradable_kitty();
+    output2.kitty_basic_data.dna = KittyDNA(H256::from_slice(b"superkalifragislisticexpialadoca"));
+    let result = TradableKittyConstraintChecker::<0>::ListKittyForSale.check(
+        &[input1.into(), input2.into()],
+        &[],
+        &[output1.clone().into(), output1.into()],
+    );
+    assert_eq!(result, Err(TradeableKittyError::DuplicateKittyFound));
+}
```

Comment:
I don't think you need a special case for colliding outputs. This should already fail because an output kitty mismatches its input kitty.

For example if I input kitty 1 and kitty 2 then output tradeable kitty 1 and tradeable kitty 1 again, then this already fails because my kitty 2 was modified in unacceptable ways.

Reply:
Actually, yeah, you are testing exactly what I mean in the test below this one.

Reply:
Fixed.